main/src/ca/uwaterloo/flix/language/phase/unification/BddFormula.scala

```diff
@@ -15,10 +15,166 @@
+
+    /**
+      * Returns `true` if `f` represents a variable.
+      */
+    override def isVar(f: BddFormula): Boolean =
+      f.dd.equalsTo(f.dd.getVariable, creator.makeFalse(), creator.makeTrue())
+
+    /**
+      * Returns a representation of TRUE.
+      */
+    override def mkTrue: BddFormula = {
+      new BddFormula(creator.makeTrue())
+    }
+
+    /**
+      * Returns a representation of FALSE.
+      */
+    override def mkFalse: BddFormula = {
+      new BddFormula(creator.makeFalse())
+    }
+
+    /**
+      * Returns a representation of the variable with the given `id`.
+      */
+    override def mkVar(id: Int): BddFormula = {
+      new BddFormula(creator.makeIthVar(id))
+    }
+
+    /**
+      * Returns a representation of the complement of `f`.
+      */
+    override def mkNot(f: BddFormula): BddFormula = {
+      new BddFormula(creator.makeNot(f.dd))
+    }
+
+    /**
+      * Returns a representation of the disjunction of `f1` and `f2`.
+      */
+    override def mkOr(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeOr(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the conjunction of `f1` and `f2`.
+      */
+    override def mkAnd(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeAnd(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns a representation of the formula `f1 == f2`.
+      */
+    override def mkXor(f1: BddFormula, f2: BddFormula): BddFormula = {
+      new BddFormula(creator.makeXor(f1.dd, f2.dd))
+    }
+
+    /**
+      * Returns the set of free variables in `f`.
+      */
+    override def freeVars(f: BddFormula): SortedSet[Int] = {
+      freeVarsAux(f.dd)
+    }
+
+    private def freeVarsAux(dd: DD): SortedSet[Int] = {
+      if (dd.isLeaf()) {
+        SortedSet.empty
+      } else {
+        SortedSet(dd.getVariable()) ++
+          freeVarsAux(dd.getLow) ++
+          freeVarsAux(dd.getHigh)
+      }
+    }
+
+    /**
+      * Applies the function `fn` to every variable in `f`.
+      */
+    override def map(f: BddFormula)(fn: Int => BddFormula): BddFormula = {
+      new BddFormula(mapAux(f.dd)(fn))
+    }
+
+    private def mapAux(dd: DD)(fn: Int => BddFormula): DD = {
+      if (dd.isLeaf()) {
+        dd
+      } else {
+        val currentVar = dd.getVariable()
+        val substDD = fn(currentVar).dd
+
+        val lowRes = mapAux(dd.getLow())(fn)
+        val highRes = mapAux(dd.getHigh())(fn)
+        creator.makeIte(substDD, highRes, lowRes)
+      }
+    }
+
+    /**
+      * Returns a representation equivalent to `f` (but potentially smaller).
+      */
+    override def minimize(f: BddFormula): BddFormula = f
+
+    /**
+      * Converts the given formula f into a type.
+      */
+    override def toType(f: BddFormula, env: Bimap[Symbol.KindedTypeVarSym, Int]): Type = {
+      createTypeFromBDDAux(f.dd, Type.True, env)
+    }
+
+    //TODO: Optimize
```

Comment:
Yes, please, if possible :)

Reply:
+ Documentation and commentary.

Reply:
The optimization is quite involved, but I have discussed with Jaco how to do it. I will leave it for now and come back to it